core/src/main/java/ai/timefold/solver/core/impl/domain/variable/custom/CustomShadowVariableDescriptor.java

```diff
@@ -49,56 +49,62 @@ public void linkVariableDescriptors(DescriptorPolicy descriptorPolicy) {
-                        + ") among planning entities in the solver configuration.");
+                throw new IllegalArgumentException(
+                        """
+                                The entityClass (%s) has a @%s annotated property (%s) with a sourceEntityClass (%s) which is not a valid planning entity.
+                                Maybe check the annotations of the class (%s).
+                                Maybe add the class (%s) among planning entities in the solver configuration."""
+                                .formatted(entityDescriptor.getEntityClass(), ShadowVariable.class.getSimpleName(),
+                                        variableMemberAccessor.getName(), sourceEntityClass, sourceEntityClass,
+                                        sourceEntityClass));
             }
         }
-        String sourceVariableName = shadowVariable.sourceVariableName();
-        VariableDescriptor<Solution_> sourceVariableDescriptor =
-                sourceEntityDescriptor.getVariableDescriptor(sourceVariableName);
+        var sourceVariableName = shadowVariable.sourceVariableName();
+        var sourceVariableDescriptor = sourceEntityDescriptor.getVariableDescriptor(sourceVariableName);
         if (sourceVariableDescriptor == null) {
-            throw new IllegalArgumentException("The entityClass (" + entityDescriptor.getEntityClass()
-                    + ") has a @" + ShadowVariable.class.getSimpleName()
-                    + " annotated property (" + variableMemberAccessor.getName()
-                    + ") with sourceVariableName (" + sourceVariableName
-                    + ") which is not a valid planning variable on entityClass ("
-                    + sourceEntityDescriptor.getEntityClass() + ").\n"
-                    + sourceEntityDescriptor.buildInvalidVariableNameExceptionMessage(sourceVariableName));
+            throw new IllegalArgumentException(
+                    """
+                            The entityClass (%s) has a @%s annotated property (%s) with sourceVariableName (%s) which is not a valid planning variable on entityClass (%s).
+                            %s"""
+                            .formatted(entityDescriptor.getEntityClass(), ShadowVariable.class.getSimpleName(),
+                                    variableMemberAccessor.getName(), sourceVariableName,
+                                    sourceEntityDescriptor.getEntityClass(),
+                                    sourceEntityDescriptor.buildInvalidVariableNameExceptionMessage(sourceVariableName)));
         }
-        Class<? extends AbstractVariableListener> variableListenerClass = shadowVariable.variableListenerClass();
+        if (!sourceVariableDescriptor.canBeUsedAsSource()) {
+            throw new IllegalArgumentException(
+                    "The entityClass (%s) has a @%s annotated property (%s) with sourceVariableName (%s) which cannot be used as source."
```

Comment:
Maybe these new exceptions should explain _why_ that's not the case. What did the user do wrong? From the message, it's not at all clear.